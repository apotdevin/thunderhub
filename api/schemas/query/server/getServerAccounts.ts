import { GraphQLList } from 'graphql';
import { ContextType } from 'api/types/apiTypes';
import { ServerAccountType } from 'api/schemas/types/QueryType';
import { SSO_ACCOUNT, SERVER_ACCOUNT } from 'src/context/AccountContext';
import { logger } from 'api/helpers/logger';
import { requestLimiter } from '../../../helpers/rateLimiter';

export const getServerAccounts = {
  type: new GraphQLList(ServerAccountType),
  resolve: async (_: undefined, params: any, context: ContextType) => {
    const { ip, accounts, account, sso, ssoVerified } = context;
    await requestLimiter(ip, 'getServerAccounts');

    const { macaroon, cert, host } = sso;
    let ssoAccount = null;
    if (macaroon && host && ssoVerified) {
      logger.debug(
        `Macaroon${
          cert ? ', certificate' : ''
        } and host (${host}) found for SSO.`
      );
      ssoAccount = {
        name: 'SSO Account',
        id: SSO_ACCOUNT,
        loggedIn: true,
        type: SSO_ACCOUNT,
      };
    }

    const currentId = account?.id;
    const withStatus =
      accounts?.map(a => ({
        ...a,
        loggedIn: a.id === currentId,
        type: SERVER_ACCOUNT,
      })) || [];

    return ssoAccount ? [...withStatus, ssoAccount] : withStatus;
  },
};
