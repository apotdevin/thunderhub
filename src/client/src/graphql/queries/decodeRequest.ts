import { gql } from '@apollo/client';

export const DECODE_REQUEST = gql`
  query DecodeRequest($request: String!) {
    decodeRequest(request: $request) {
      chain_address
      cltv_delta
      description
      description_hash
      destination
      destination_node {
        node {
          alias
        }
      }
      expires_at
      id
      routes {
        base_fee_mtokens
        channel
        cltv_delta
        fee_rate
        public_key
      }
      tokens
    }
  }
`;
