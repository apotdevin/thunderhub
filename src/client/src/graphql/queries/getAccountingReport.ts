import { gql } from '@apollo/client';

export const GET_ACCOUNTING_REPORT = gql`
  query GetAccountingReport(
    $category: String
    $currency: String
    $fiat: String
    $month: String
    $year: String
  ) {
    getAccountingReport(
      category: $category
      currency: $currency
      fiat: $fiat
      month: $month
      year: $year
    )
  }
`;
