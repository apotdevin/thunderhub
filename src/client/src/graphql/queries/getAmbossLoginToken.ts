import { gql } from '@apollo/client';

export const GET_AMBOSS_LOGIN_TOKEN = gql`
  query GetAmbossLoginToken($redirect_url: String) {
    getAmbossLoginToken(redirect_url: $redirect_url)
  }
`;
