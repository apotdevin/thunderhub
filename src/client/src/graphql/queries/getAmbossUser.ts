import { gql } from '@apollo/client';

export const GET_AMBOSS_USER = gql`
  query GetAmbossUser {
    getAmbossUser {
      subscription {
        end_date
        subscribed
        upgradable
      }
      backups {
        last_update
        last_update_size
        total_size_saved
        available_size
        remaining_size
      }
      ghost {
        username
      }
    }
  }
`;
