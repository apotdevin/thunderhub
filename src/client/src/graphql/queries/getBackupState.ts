import { gql } from '@apollo/client';

export const GET_BACKUP_STATE = gql`
  query GetBackupState {
    getBackupState
  }
`;
