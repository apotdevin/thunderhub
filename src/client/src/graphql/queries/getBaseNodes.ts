import { gql } from '@apollo/client';

export const GET_BASE_NODES = gql`
  query GetBaseNodes {
    getBaseNodes {
      _id
      name
      public_key
      socket
    }
  }
`;
