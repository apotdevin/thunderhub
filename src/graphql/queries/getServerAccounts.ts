import { gql } from '@apollo/client';

export const GET_SERVER_ACCOUNTS = gql`
  query GetServerAccounts {
    getServerAccounts {
      name
      id
      loggedIn
      type
    }
  }
`;
