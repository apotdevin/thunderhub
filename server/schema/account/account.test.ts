import testServer from 'server/tests/testServer';
import { GET_SERVER_ACCOUNTS } from 'src/graphql/queries/getServerAccounts';
import { ContextMockNoSSO } from 'server/tests/testMocks';

jest.mock('ln-service');

describe('Account Resolvers', () => {
  describe('getServerAccounts', () => {
    test('full accounts', async () => {
      const { query } = testServer();

      const res = await query({
        query: GET_SERVER_ACCOUNTS,
      });

      expect(res.errors).toBe(undefined);
      expect(res).toMatchSnapshot();
    });
    test('without SSO', async () => {
      const { query } = testServer(ContextMockNoSSO);

      const res = await query({
        query: GET_SERVER_ACCOUNTS,
      });

      expect(res.errors).toBe(undefined);
      expect(res).toMatchSnapshot();
    });
    test('without accounts', async () => {
      const { query } = testServer(ContextMockNoSSO);

      const res = await query({
        query: GET_SERVER_ACCOUNTS,
      });

      expect(res.errors).toBe(undefined);
      expect(res).toMatchSnapshot();
    });
  });
});
