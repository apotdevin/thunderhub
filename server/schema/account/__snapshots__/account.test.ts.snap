// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Resolvers getServerAccounts full accounts 1`] = `
Object {
  "data": Object {
    "getServerAccounts": Array [
      Object {
        "id": "accountID",
        "loggedIn": false,
        "name": "account",
        "type": "server",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Account Resolvers getServerAccounts without SSO 1`] = `
Object {
  "data": Object {
    "getServerAccounts": Array [
      Object {
        "id": "accountID",
        "loggedIn": false,
        "name": "account",
        "type": "server",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Account Resolvers getServerAccounts without accounts 1`] = `
Object {
  "data": Object {
    "getServerAccounts": Array [
      Object {
        "id": "accountID",
        "loggedIn": false,
        "name": "account",
        "type": "server",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
